File: backend/services/google-earth/fire/fireProtection.js
import path from "path";
import { spawn } from "child_process";
import { fileURLToPath } from "url";
import fs from "fs";

// --- Calculate __dirname equivalent in ESM ---
const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

/**
 * Executes the Python GEE fire detection script.
 * @param {Object} regionGeoJson - GeoJSON object for the region to analyze
 * @param {string} regionId - Identifier for the region
 * @param {string} credentialsPath - Path to GCP credentials file
 * @param {number} [daysBack] - Optional: how many days back should the fire detection look
 * @returns {Promise<Object>} - Analysis results
 */
function runFireProtectionCheck(
  regionGeoJson,
  regionId,
  credentialsPath,
  daysBack
) {
  return new Promise((resolve, reject) => {
    const pythonExecutable = "python";
    const scriptFilename = "fire_protection.py";
    const scriptPath = path.resolve(__dirname, scriptFilename);

    if (!fs.existsSync(scriptPath)) {
      return reject(
        new Error(`Python script not found at path: ${scriptPath}`)
      );
    }

    console.log(`Executing Python script: ${scriptPath}`);
    console.log(`For region: ${regionId}`);
    console.log(`Passing credentials path: ${credentialsPath}`);

    const pythonProcess = spawn(pythonExecutable, [
      scriptPath,
      credentialsPath,
    ]);

    const inputData = {
      geometry: regionGeoJson,
      region_id: regionId,
    };

    if (daysBack !== undefined && daysBack !== null) {
      inputData.days_back = daysBack;
    }

    const inputJsonString = JSON.stringify(inputData);

    let scriptOutput = "";
    let scriptError = "";

    pythonProcess.stdout.on("data", (data) => {
      scriptOutput += data.toString();
    });

    pythonProcess.stderr.on("data", (data) => {
      scriptError += data.toString();
      console.error(`Python stderr: ${data}`);
    });

    pythonProcess.on("close", (code) => {
      console.log(`Python script exited with code ${code}`);
      if (code === 0) {
        try {
          const trimmedOutput = scriptOutput.trim();

          if (!trimmedOutput) {
            return reject(new Error("Python script returned empty output"));
          }

          const result = JSON.parse(trimmedOutput);
          console.log("Successfully parsed Python output.");
          resolve(result);
        } catch (parseError) {
          console.error("Failed to parse Python JSON output:", parseError);
          console.error("Raw Python output:", scriptOutput);
          reject(
            new Error(`Failed to parse JSON output: ${parseError.message}`)
          );
        }
      } else {
        console.error(`Python script failed with exit code ${code}`);
        reject(
          new Error(
            `Python script failed with code ${code}. Error output: ${scriptError}`
          )
        );
      }
    });

    pythonProcess.on("error", (err) => {
      console.error("Failed to start Python subprocess:", err);
      reject(new Error(`Failed to start Python process: ${err.message}`));
    });

    try {
      console.log("Writing input data to Python stdin:", inputJsonString);
      pythonProcess.stdin.write(inputJsonString);
      pythonProcess.stdin.end();
    } catch (stdinError) {
      console.error("Error writing to Python stdin:", stdinError);
      reject(new Error(`Error writing to Python stdin: ${stdinError.message}`));
    }
  });
}

export { runFireProtectionCheck };

File: backend/services/google-earth/fire/fireProtection.py
import sys
import json
import os
import ee
import datetime
import time
import traceback

DEFAULT_DAYS_BACK = 5  # How many days back to check for fires

MODIS_FIRE_COLLECTION = 'MODIS/006/MCD14DL'  # NASA MODIS Fire/Hotspot Thermal Anomalies
gcp_project_id = 'project-ultron-457221'

def initialize_gee(credentials_path_arg):
    global gcp_project_id
    try:
        credentials_path = credentials_path_arg
        print(f"DEBUG: Received credentials path via argument: {credentials_path}", file=sys.stderr)
        if not credentials_path or not os.path.exists(credentials_path):
            print(f"ERROR: Credentials file not found or path empty: {credentials_path}", file=sys.stderr)
            return False
        credentials = ee.ServiceAccountCredentials(None, key_file=credentials_path)
        ee.Initialize(credentials=credentials, project=gcp_project_id, opt_url='https://earthengine-highvolume.googleapis.com')
        print(f"GEE Initialized OK for project: {gcp_project_id}.", file=sys.stderr)
        return True
    except ee.EEException as e:
        print(f"ERROR: Failed GEE init: {e}", file=sys.stderr)
        return False
    except Exception as e:
        print(f"ERROR: Unexpected GEE init error: {e}", file=sys.stderr)
        print(traceback.format_exc(), file=sys.stderr)
        return False

def detect_active_fires(region_geometry, days_back):
    try:
        now = datetime.datetime.now(datetime.timezone.utc)
        end_date = ee.Date(now)
        start_date = end_date.advance(-days_back, 'day')

        fire_collection = ee.FeatureCollection(MODIS_FIRE_COLLECTION) \
            .filterDate(start_date, end_date) \
            .filterBounds(region_geometry)

        fire_count = fire_collection.size().getInfo()
        print(f"Detected {fire_count} active fire pixels in region (last {days_back} days)", file=sys.stderr)

        fires_list = []
        if fire_count > 0:
            # For illustration, get up to 20 fire points' info
            sample = fire_collection.limit(20).getInfo()
            for feat in sample['features']:
                fire_info = {
                    "acq_date": feat['properties'].get('acq_date'),
                    "acq_time": feat['properties'].get('acq_time'),
                    "brightness": feat['properties'].get('brightness'),
                    "confidence": feat['properties'].get('confidence'),
                    "latitude": feat['geometry']['coordinates'][1],
                    "longitude": feat['geometry']['coordinates'][0]
                }
                fires_list.append(fire_info)

        return {
            "status": "success",
            "active_fire_count": fire_count,
            "fires": fires_list,
            "days_back": days_back
        }
    except ee.EEException as gee_error:
        error_str = str(gee_error)
        print(f"ERROR: GEE computation failed: {error_str}", file=sys.stderr)
        return {
            "status": "error",
            "message": f"GEE Computation Error: {error_str}"
        }
    except Exception as e:
        print(f"ERROR: Unexpected Python error during GEE processing: {e}", file=sys.stderr)
        print(traceback.format_exc(), file=sys.stderr)
        return {
            "status": "error",
            "message": f"Python Script Error: {e}"
        }

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("ERROR: Missing credentials file path argument.", file=sys.stderr)
        print(json.dumps({"status": "error", "message": "Missing credentials file path argument."}))
        sys.exit(1)
    credentials_path_from_arg = sys.argv[1]

    input_data_str = sys.stdin.read()
    days_back = DEFAULT_DAYS_BACK
    region_id = "unknown_region"

    try:
        input_params = json.loads(input_data_str)
        geojson_geometry = input_params['geometry']
        days_back = int(input_params.get('days_back', DEFAULT_DAYS_BACK))
        region_id = str(input_params.get('region_id', region_id))
        print(f"Received job: region='{region_id}', days_back={days_back}", file=sys.stderr)
    except Exception as e:
        print(f"ERROR: Invalid stdin params: {e}", file=sys.stderr)
        print(json.dumps({"status": "error", "message": f"Invalid Stdin Param: {e}", "region_id": region_id}))
        sys.exit(1)

    if not initialize_gee(credentials_path_from_arg):
        print(json.dumps({"status": "error", "message": "GEE initialization failed.", "region_id": region_id}))
        sys.exit(1)

    ee_geometry = None
    try:
        geom_type = geojson_geometry.get('type')
        coords = geojson_geometry.get('coordinates')
        if geom_type == 'Polygon':
            ee_geometry = ee.Geometry.Polygon(coords)
        elif geom_type == 'MultiPolygon':
            ee_geometry = ee.Geometry.MultiPolygon(coords)
        elif geom_type == 'Point':
            ee_geometry = ee.Geometry.Point(coords).buffer(10000)  # 10km buffer for points
        else:
            raise ValueError(f"Unsupported type: {geom_type}")
    except Exception as e:
        print(f"ERROR: GeoJSON convert fail: {e}", file=sys.stderr)
        print(json.dumps({"status": "error", "message": f"GeoJSON Error: {e}", "region_id": region_id}))
        sys.exit(1)

    print(f"Starting GEE fire detection analysis for region: {region_id}...", file=sys.stderr)
    start_time = time.time()
    analysis_result = detect_active_fires(ee_geometry, days_back)
    end_time = time.time()
    print(f"GEE analysis duration: {end_time - start_time:.2f} seconds.", file=sys.stderr)
    analysis_result['region_id'] = region_id
    print(json.dumps(analysis_result))
    if analysis_result.get("status") == "success":
        sys.exit(0)
    else:
        sys.exit(1)